### DISCUSSION TITLE
How to quickly expand a list horizontally into multiple columns?

### DISCUSSION THREAD

**Discussion by tt0203** [duckdb/duckdb#11337] [Ideas]
There is a list column of 5 integers in my tb1 table.
create table tb1(x varchar,y integer[5]);
insert into tb1 values('a',[0,1,2,3,4]),('b',[5,6,7,8,9]);
select y[1] y1,y[2] y2,y[3] y3,y[4] y4,y[5] y5 from tb1;
How to quickly expand a list horizontally into multiple columns?

**Comment by Tishj** [comment]
https://duckdb.org/docs/sql/query_syntax/unnest.html

**Reply by tt0203** [reply]
Although it can be achieved by borrowing the row position to mark the pivot column after vertical expansion, I feel it is still too troublesome. Why not make a function that automatically expands horizontally?For example, a column of a text string needs to be split into multiple columns according to the delimiter,,,

**Comment by cmdlineluser** [comment]
You can *"wildcard"* select from a [Struct](https://duckdb.org/docs/sql/data_types/struct.html)

```python
duckdb.sql("""
with t as (
   from tb1
   select {'y1': y[1], 'y2': y[2], 'y3': y[3], 'y4': y[4], 'y5': y[5]} as y
)
from t select y.*
""")
# ┌───────┬───────┬───────┬───────┬───────┐
# │  y1   │  y2   │  y3   │  y4   │  y5   │
# │ int32 │ int32 │ int32 │ int32 │ int32 │
# ├───────┼───────┼───────┼───────┼───────┤
# │     0 │     1 │     2 │     3 │     4 │
# │     5 │     6 │     7 │     8 │     9 │
# └───────┴───────┴───────┴───────┴───────┘
```

A Struct will also UNNEST horizontally:

```python
duckdb.sql("""
from tb1
select * replace (
   unnest(
      {'y1': y[1], 'y2': y[2], 'y3': y[3], 'y4': y[4], 'y5': y[5]}
   ) as y
)
""")
# ┌─────────┬───────┬───────┬───────┬───────┬───────┐
# │    x    │  y1   │  y2   │  y3   │  y4   │  y5   │
# │ varchar │ int32 │ int32 │ int32 │ int32 │ int32 │
# ├─────────┼───────┼───────┼───────┼───────┼───────┤
# │ a       │     0 │     1 │     2 │     3 │     4 │
# │ b       │     5 │     6 │     7 │     8 │     9 │
# └─────────┴───────┴───────┴───────┴───────┴───────┘
```

However, there doesn't appear to be a way to create a Struct "dynamically", only a [Map](https://duckdb.org/docs/sql/data_types/map.html)

```python
duckdb.sql("""
from tb1 
select * replace (
   map(
      [ 'y' || n + 1 for n in range(len(y)) ], 
      [ y[n + 1] for n in range(len(y)) ]
   ) as y
)
""")
# ┌─────────┬────────────────────────────────┐
# │    x    │               y                │
# │ varchar │     map(varchar, integer)      │
# ├─────────┼────────────────────────────────┤
# │ a       │ {y1=0, y2=1, y3=2, y4=3, y5=4} │
# │ b       │ {y1=5, y2=6, y3=7, y4=8, y5=9} │
# └─────────┴────────────────────────────────┘
```

But UNNEST is not allowed on a Map, and we cannot create a Struct from one.

It seems like what is missing is some sort of `list_to_struct(list_name, column_name_format_string)` function which lets you create a struct from a list/array?

Then we could simply use  `unnest(list_to_struct(y))` instead of the current `pivot` workaround.

**Reply by tt0203** [reply]
What you said is great, that’s what I thought, thank you.
